stdlib/public/TensorFlow/Ops.swift

```diff
@@ -1597,52 +1597,257 @@ public extension Tensor where Scalar : Numeric {
 // Indexing and slicing
 //===----------------------------------------------------------------------===//
 
-public extension Tensor {
-  /// Access the element tensor specified by an index in the leading dimension.
-  /// - Parameter index: Index of the element tensor.
-  @inlinable
-  subscript(index: Int32) -> Tensor {
-    get {
-      let slice = Raw.stridedSlice(
-        self, begin: Tensor<Int32>([index]), end: Tensor<Int32>([index + 1]),
-        strides: Tensor<Int32>([1]))
-      return slice.squeezingShape(at: 0)
-    }
-    set {
-      let leftElements = self[0..<index]
-      let rightElements = self[index+1..<shape[0]]
-      self = Raw.concatV2([leftElements, newValue.rankLifted(), rightElements],
-                          axis: Tensor<Int32>(0))
-    }
-  }
-
-  /// Access the subtensor specified by a contiguous range of indices.
-  /// - Parameter bounds: Contiguous range of indices.
-  @inlinable
-  subscript(bounds: Range<Int32>) -> Tensor {
-    return Raw.stridedSlice(
-      self, begin: Tensor<Int32>([bounds.lowerBound]),
-      end: Tensor<Int32>([bounds.upperBound]), strides: Tensor<Int32>([1]))
-  }
-
-  // TODO(danielzheng): Add strided slices? (increment by something different
-  // than 1)
-  // Ideas for strided slice API: it could be another subscript method, or it
-  // be a top level `stride` function like Swift's `stride(from:to:by:)`.
+// TODO: Negative indexing and strides syntax.
 
+public extension Tensor {
   /// Extracts a slice from the tensor defined by lower and upper bounds for
   /// each dimension.
   ///
   /// - Parameter lowerBounds: The lower bounds at each dimension.
   /// - Parameter upperBounds: The upper bounds at each dimension.
   @inlinable @inline(__always)
+  @differentiable(wrt: self)
   func slice(lowerBounds: [Int32], upperBounds: [Int32]) -> Tensor {
     /// TODO: Precondition `lowerBounds.count == upperBounds.count`,
     /// preferably in graph.
     let lowerBoundsTensor = Tensor<Int32>(lowerBounds)
-    return Raw.slice(
-      self,
+    return self.slice(
```

Comment:
No need for explicit `self.`.

```suggestion
    return slice(
```